firecracker-control/local.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package service
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math"
+	"math/big"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/plugin"
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+)
+
+var (
+	_ proto.FirecrackerServer = (*local)(nil)
+)
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type: plugin.ServicePlugin,
+		ID:   localPluginID,
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			log.G(ic.Context).Debugf("initializing %s plugin (root: %q)", localPluginID, ic.Root)
+			return newLocal(ic)
+		},
+	})
+}
+
+// TODO: Remove this interface after merging PR #91
+// See https://github.com/firecracker-microvm/firecracker-go-sdk/pull/91
+type machineIface interface {
+	Start(context.Context) error
+	StopVMM() error
+	Wait(context.Context) error
+	SetMetadata(context.Context, interface{}) error
+}
+
+type instance struct {
+	cfg     *firecracker.Config
+	machine machineIface
+}
+
+type local struct {
+	vm          map[string]*instance
+	vmLock      sync.RWMutex
+	rootPath    string
+	findVsockFn func(context.Context) (uint32, error)
+}
+
+func newLocal(ic *plugin.InitContext) (*local, error) {
+	if err := os.MkdirAll(ic.Root, 0750); err != nil && !os.IsExist(err) {
+		return nil, errors.Wrapf(err, "failed to create root directory: %s", ic.Root)
+	}
+
+	return &local{
+		vm:          make(map[string]*instance),
+		rootPath:    ic.Root,
+		findVsockFn: findNextAvailableVsockCID,
+	}, nil
+}
+
+// CreateVM creates new Firecracker VM instance
+func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*proto.CreateVMResponse, error) {
+	id, err := s.makeID()
```

Comment:
I wouldn't require it from client unless we have use cases when this is useful (I can't imaging any for now).

Reply:
It could be useful for idempotency; i.e. the client wants a request to fail if a VM with that ID already exists (and do so in a way free from race conditions)

Reply:
Discussed offline. Will get rid of id management in subsequent PRs.